blast/ncbi/doc/man/fastacmd.1
.TH FASTACMD 1 2003-11-10 NCBI "NCBI Tools User's Manual"
.SH NAME
fastacmd \- retrieve FASTA sequences from a BLAST database
.SH SYNOPSIS
.B fastacmd
[\|\fB\-\fP\|]
[\|\fB\-D\fP\|]
[\|\fB\-I\fP\|]
[\|\fB\-L\fP\ \fIstart\fP,\fIstop\fP\|]
[\|\fB\-P\fP\ \fIN\fP\|]
[\|\fB\-S\fP\ \fIN\fP\|]
[\|\fB\-T\fP\|]
[\|\fB\-a\fP\|]
[\|\fB\-c\fP\|]
[\|\fB\-d\fP\ \fIstr\fP\|]
[\|\fB\-i\fP\ \fIstr\fP\|]
[\|\fB\-l\fP\ \fIN\fP\|]
[\|\fB\-o\fP\ \fIfilename\fP\|]
[\|\fB\-p\fP\ \fItype\fP\|]
[\|\fB\-s\fP\ \fIstr\fP\|]
[\|\fB\-t\fP\|]
.SH DESCRIPTION
\fBfastacmd\fP retrieves FASTA formatted sequences from a
\fBblast\fP(1) database formatted using the '\fB\-o\fP' option.  An
example \fBfastacmd\fP call would be
.PP
.ce
fastacmd -d nr -s p38398
.SH OPTIONS
A summary of options is included below.
.TP
\fB\-\fP
Print usage message
.TP
\fB\-D\fP
Dump the entire database in FASTA format
.TP
\fB\-I\fP
Print database information only (overrides all other options)
.TP
\fB\-L\fP\ \fIstart\fP,\fIstop\fP
Range of sequence to extract (0 in start is beginning of sequence, 0
in stop is end of sequence, default is whole sequence)
.TP
\fB\-P\fP\ \fIN\fP
Retrieve sequences with Protein Identification Group (PIG) \fIN\fP.
.TP
\fB\-S\fP\ \fIN\fP
Strand on subsequence (nucleotide only):
.RS
.PD 0
.IP 1
top (default)
.IP 2
bottom
.PD
.RE
.TP
\fB\-T\fP
Print taxonomic information for requested sequence(s)
.TP
\fB\-a\fP
Retrieve duplicate accessions
.TP
\fB\-c\fP
Use ^A (\e001) as non-redundant defline separator
.TP
\fB\-d\fP\ \fIstr\fP
Database (default is nr)
.TP
\fB\-i\fP\ \fIstr\fP
Input file wilth GIs/accessions/locuses for batch retrieval
.TP
\fB\-l\fP\ \fIN\fP
Line length for sequence (default = 80)
.TP
\fB\-o\fP\ \fIfilename\fP
Output file (default = stdout)
.TP
\fB\-p\fP\ \fItype\fP
Type of file:
.RS
.PD 0
.IP G
guess (default) - look for protein, then nucleotide
.IP T
protein
.IP F
nucleotide
.PD
.RE
.TP
\fB\-s\fP\ \fIstr\fP
Search string: GIs, accessions and locuses may be used delimited by
commas
.TP
\fB\-t\fP
Definition line should contain target GI only
.SH EXIT STATUS
.RS
.PD 0
.IP 0
Completed successfully.
.IP 1
An error (other than those below) occurred.
.IP 2
The BLAST database was not found.
.IP 3
A search (accession, GI, or taxonomy info) failed.
.IP 4
No taxonomy database was found.
.PD
.RE
.SH AUTHOR
The National Center for Biotechnology Information.
.SH SEE ALSO
.BR blast (1),
fastacmd.txt


QuickTimeStreamingServer/ReleaseNotes.txt

                                 RELEASE NOTES
                             Darwin Streaming Server 

These release notes are for the current development version of Darwin
Streaming Server.

------------------------------------------------------------------------
Date: 8/7/2001
Version: 3.0.1+development [v349]
cvs tag: qtssServer-349

*   Updated build platforms: Solaris 8, Redhat Linux 7.1, FreeBSD 4.3, 
    Mac OS X 10.1
*   Retransmit/overbuffering (Skip Protection) algorithms tuned
    substantially
*   "reliable_udp" attribute defaults to "enabled" 
*   Various WebAdmin fixes for Windows platform

------------------------------------------------------------------------
Date: 8/9/2001
Version: 3.0.1+development [v350]
cvs tag: qtssServer-350

*   Minor bug fixes.

------------------------------------------------------------------------
Date: 8/14/2001
Version: 3.0.1+development [v351]
cvs tag: qtssServer-351

*   QTSSFileModule - enabled caching based on preferences (shared and
    private).
*   Add buffer delay to on demand movies.
*   QTFile - code cleanup.
*   Changed some paths for the installer.
*   Added support for MP3Broadcaster on Windows.

------------------------------------------------------------------------
Date: 8/16/2001
Version: 3.0.1+development [v352]
cvs tag: qtssServer-352

*   Added MP3StreamingModule to all builds.
*   Fixed gdb script to execute QTSS from build dir on MacOSX.

------------------------------------------------------------------------
Date: 8/19/2001
Version: 3.0.1+development [v353]
cvs tag: qtssServer-353

*   Added:
        add_seconds_to_client_buffer_delay -- Increases client buffer
                                delay before playing movie.
        record_movie_file_sdp -- Records a .sdp file for a .mov file.
                                Appears as moviename.mov.sdp
        enable_movie_file_sdp -- Plays a .mov.sdp file for a .mov movie.
*   Fixed server to play .mov.sdp files.
*   QTSSFileModule enables both private and shared movie cache buffers.

------------------------------------------------------------------------
Date: 8/23/2001
Version: 3.0.1+development [v354]
cvs tag: qtssServer-354

*   Added prefs for MP3StreamingModule.
*   Default Reliable UDP dir is / on UNIX and \ on NT
*   Use Enable Reliable UDP pref to disable Reliable UDP
*   Included new Web Admin
*   Changed relay preferences to XML format.
        Old relay config files will not work.

------------------------------------------------------------------------
Date: 08/28/2001
Version: 3.0.1+development [v355]
cvs tag: qtssServer-355

*	Added preference:
	 RTSP_error_message to enable debugging error strings returned in 
	 the content body of RTSP error responses.
	
*	Changed RTSP code to:
	1) require a Cseq field for OPTIONS requests
	2) allow missing Range headers in a PLAY request (defaults to entire movie)
	3) add content length header to Error responses containing error messages
	4) add SDP o= and t= fields to on-demand movie Describe responses

*      Changed Relay to relay incoming ANNOUNCEd broadcasts
       1) All ANNOUNCEd broadcasts, or
       2) All broadcasts from an IP address
       3) A broadcast URL from an IP address

       (a) Can still configure the relay to relay UDP streams
       (b) Can pull an SDP or movie from source and relay it

*      Changed Relay to ANNOUNCE the SDPs to its destinations
       1) for ANNOUNCED sources
       2) for sources that are pulled from the source server

*      Changed Relay config file to be in XML format
       1) relayconfig.xml-Sample will be installed instead of old streamingrelay.conf

*	Added General Settings and Change Password to UI
	1) Require existing password to change to new
	2) Require the user to enter the same password twice
	3) Change the auth cookie so the user won't have to re-authenticate

-------------------------------------------------------------------------------------------------
Date: 8/31/2001
Version: 3.0.1+development [v356]
cvs tag: qtssServer-356

*	 Admin Module change
	 1) Added an admin module pref called "AdministratorGroup" with
	 a default value of "admin"
	 2) Changed Admin Module to allow all users that belong to the
	 AdministratorGroup, instead of allowing only one user called
	 "streamingadmin".
	 3) A default qtgroups gets installed from now. The default qtusers
	 file has user 'streamingadmin' with password 'default'. The default
	 qtgroups file has the group 'admin' with the user 'streamingadmin'.
	 4) The login name for the web admin can be changed by adding the user
	 to the qtusers file and adding the user to the 'admin' group in the
	 qtgroups file.

*	Added "RTSP_debug_printfs" server pref to show the incoming and outgoing RTSP.
	Fixed Session header response to remove ";timeout=" field if the value is not defined.
	Fixed .mov.sdp Describe response to include the Last-Modified and Cache-Control headers.
	Fixed RTSP Session header parser to also stop on a ';'
	Changes shared file buffer size to default to maximum size if the bit-rate is 0 and no fixed 
	buffer size is defined.

-------------------------------------------------------------------------------------------------
Date: 9/04/2001
Version: 3.0.1+development [v357]
cvs tag: qtssServer-357

*	more implementation code for Web admin added.
*	New XML parser for reading prefs.
*	full implementation of MP3 streaming module is included.

-------------------------------------------------------------------------------------------------
Date: 9/06/2001
Version: 3.0.1+development [v359]
cvs tag: qtssServer-359

*	Relay that pulls source from the same server can be setup now
	[rtsp_source SOURCE with own IP address as the source_addr]
   
-------------------------------------------------------------------------------------------------
Date: 9/10/2001
Version: 3.0.1+development [v361]
cvs tag: qtssServer-361

*	UI: play and stop buttons in playlist screen now start/stop playlists immediately

blast/ncbi/demo/fastacmd.c
static char const rcsid[] = "$Id: fastacmd.c,v 6.34 2004/12/04 03:39:48 camacho Exp $";

/* $Id: fastacmd.c,v 6.34 2004/12/04 03:39:48 camacho Exp $
* ===========================================================================
*
*                            PUBLIC DOMAIN NOTICE
*               National Center for Biotechnology Information
*
*  This software/database is a "United States Government Work" under the
*  terms of the United States Copyright Act.  It was written as part of
*  the author's official duties as a United States Government employee and
*  thus cannot be copyrighted.  This software/database is freely available
*  to the public for use. The National Library of Medicine and the U.S.
*  Government have not placed any restriction on its use or reproduction.
*
*  Although all reasonable efforts have been taken to ensure the accuracy
*  and reliability of the software and data, the NLM and the U.S.
*  Government do not and cannot warrant the performance or results that
*  may be obtained by using this software or data. The NLM and the U.S.
*  Government disclaim all warranties, express or implied, including
*  warranties of performance, merchantability or fitness for any particular
*  purpose.
*
*  Please cite the author in any work or product based on this material.
*
* ===========================================================================
*
* File Name:  $RCSfile: fastacmd.c,v $
*
* Author:  Sergei Shavirin
*
* Initial Version Creation Date: 05/20/1997
*
* $Revision: 6.34 $
*
* File Description:
*        FASTA retrievel system using ISAM indexes
*
* $Log: fastacmd.c,v $
* Revision 6.34  2004/12/04 03:39:48  camacho
* Set range of data on -D option
*
* Revision 6.33  2004/12/03 04:58:06  camacho
* Fix name conflict in enumeration for fastacmd dump types
*
* Revision 6.32  2004/12/02 20:37:40  camacho
* + fastacmd feature to dump list of gis
*
* Revision 6.31  2004/06/30 19:52:00  camacho
* Added #include <blfmtutl.h>
*
* Revision 6.30  2004/05/13 20:54:45  coulouri
* spell 'loci' correctly
*
* Revision 6.29  2003/05/30 17:31:09  coulouri
* add rcsid
*
* Revision 6.28  2003/04/15 19:09:41  camacho
* Added option to retrieve sequences by PIG
*
* Revision 6.27  2002/11/21 21:35:54  camacho
* Make sure the proper exit code is returned
*
* Revision 6.26  2002/08/12 12:40:55  camacho
* Fix for unresolved symbol in Win32 build
*
* Revision 6.25  2002/08/09 19:41:25  camacho
* 1) Added blast version number to command-line options
* 2) Added explanations for some default parameters
*
* Revision 6.24  2002/07/30 21:02:17  camacho
* Added explanation for -T option
*
* Revision 6.23  2002/07/30 15:30:47  camacho
* 1. Added explanation for -L option
* 2. Moved function to parse SeqLocs to readdb.c
*
* Revision 6.22  2002/07/18 22:17:49  madden
* Revert last change
*
* Revision 6.21  2002/07/18 18:49:43  madden
* Set SeqLoc to NULL always
*
* Revision 6.20  2002/07/14 21:02:08  camacho
* Added extra features to fastacmd
*
* Revision 6.19  2002/05/02 21:59:31  camacho
* Clarified database parameter default
*
* Revision 6.18  2001/12/18 13:01:52  camacho
* Added new flag -D to dump blast database in FASTA format
*
* Revision 6.17  2001/12/10 19:17:32  camacho
* Added option to allow fastacmd to use Ctrl-As as defline separators.
*
* Revision 6.16  2001/10/19 19:46:26  camacho
* Added new feature to dump FASTA files from blast databases, added support for the new database format
*
* Revision 6.15  2000/10/16 20:47:35  madden
* Add -o option to write output to file
*
* Revision 6.14  2000/06/28 16:56:52  madden
* Call Fastacmd_Search_ex, Boolean for target gi only
*
* Revision 6.13  2000/03/08 15:26:03  madden
* Add return statement to Main fct, purify nit
*
* Revision 6.12  2000/01/12 21:05:00  egorov
* Use Fastacmd API
*
* Revision 6.11  1999/12/21 21:26:22  egorov
* Use new parameter of readdb_gi2seq function
*
* Revision 6.10  1999/12/17 20:48:54  egorov
* Fix 'gcc -Wall' warnings and remove old stuff.
*
* Revision 6.9  1999/09/28 19:02:47  egorov
* In the new version of the 'fastacmd' there is not need to
* specify database name if search a GI.
*
* Revision 6.8  1999/02/23 17:17:32  madden
* Remove unused static _accession functions
*
* Revision 6.7  1998/02/11 18:06:43  madden
* Fix for reading IDs in from a file
*
* Revision 6.6  1998/02/06 18:26:35  madden
* Removed stripping of white spaces
*
* Revision 6.5  1998/01/29 19:47:02  madden
* Changed second call from BioseqRawToFasta to BioseqRawToFastaExtra
*
* Revision 6.4  1998/01/27 20:27:03  madden
* Added option to specify sequence line length
*
* Revision 6.3  1998/01/23 21:56:16  madden
* Error messages sent to stderr
*
* Revision 6.2  1998/01/16 22:04:20  madden
* Call to readdb_new_ex, fixed FUM
*
* Revision 6.1  1997/11/07 16:19:15  shavirin
* Added possibility to retrieve redundant accessions
*
* Revision 6.0  1997/08/25 18:19:56  madden
* Revision changed to 6.0
*
* Revision 5.2  1997/05/20 21:00:45  shavirin
* Remove spurious error message in accession/locus retrieval
*
* Revision 5.1  1997/05/20 15:47:30  shavirin
* Initial revision
*
*
* ==========================================================================
*/

#include <ncbi.h>
#include <objseq.h>
#include <objsset.h>
#include <sequtil.h>
#include <seqport.h>
#include <tofasta.h>
#include <readdb.h>
#include <blast.h>
#include <blfmtutl.h>

static Args myargs [] = {
    { "Database",                                               /* 0 */
      "nr", NULL, NULL, TRUE, 'd', ARG_STRING, 0.0, 0, NULL},
    { "Type of file\n"                                          /* 1 */
      "         G - guess mode (look for protein, then nucleotide)\n"
      "         T - protein   \n"
      "         F - nucleotide", 
      "G", NULL,NULL,TRUE,'p',ARG_STRING,0.0,0,NULL},
    { "Search string: GIs, accessions and loci may be used delimited\n"
      "      by comma.",                                        /* 2 */
      NULL, NULL, NULL, TRUE, 's', ARG_STRING, 0.0, 0, NULL},
    { "Input file wilth GIs/accessions/loci for batch\n"
      "      retrieval",/* 3 */
      NULL, NULL, NULL, TRUE, 'i', ARG_STRING, 0.0, 0, NULL},
    { "Retrieve duplicate accessions",                          /* 4 */
      "F", NULL, NULL, TRUE, 'a', ARG_BOOLEAN, 0.0, 0, NULL},
    { "Line length for sequence",                               /* 5 */
      "80", NULL, NULL, TRUE, 'l', ARG_INT, 0.0, 0, NULL},
    { "Definition line should contain target gi only",          /* 6 */
      "F", NULL, NULL, TRUE, 't', ARG_BOOLEAN, 0.0, 0, NULL},
    { "Output file",                                            /* 7 */ 
      "stdout", NULL, NULL, TRUE, 'o', ARG_FILE_OUT, 0.0, 0, NULL},
    { "Use Ctrl-A's as non-redundant defline separator",        /* 8 */
      "F", NULL, NULL, TRUE, 'c', ARG_BOOLEAN, 0.0, 0, NULL},
    { "Dump the entire database as (default is not to dump anything):\n"
      "      1 FASTA\n"
      "      2 Gi list\n",                                      /* 9 */
      "0", "0", "2", TRUE, 'D', ARG_INT, 0.0, 0, NULL},
    { "Range of sequence to extract (Format: start,stop)\n"
      "      0 in 'start' refers to the beginning of the sequence\n"
      "      0 in 'stop' refers to the end of the sequence",
      "0,0", NULL, NULL, TRUE, 'L', ARG_STRING, 0.0, 0, NULL},  /* 10 */
    { "Strand on subsequence (nucleotide only): 1 is top, 2 is bottom",
      "1", NULL, NULL, FALSE, 'S', ARG_INT, 0.0, 0, NULL},      /* 11 */
    { "Print taxonomic information for requested sequence(s)",
      "F", NULL, NULL, FALSE, 'T', ARG_BOOLEAN, 0.0, 0, NULL},  /* 12 */
    { "Print database information only (overrides all other options)",
      "F", NULL, NULL, FALSE, 'I', ARG_BOOLEAN, 0.0, 0, NULL},  /* 13 */
    { "Retrieve sequences with this PIG",
      NULL, NULL, NULL, TRUE, 'P', ARG_INT, 0.0, 0, NULL},     /* 14 */
};

Int2 Main (void)
{
    CharPtr	database, searchstr, batchfile;
    Int4	linelen, pig;
    Boolean	dupl, target, use_ctrlAs, taxonomy_info, dbinfo_only;
    EBlastDbDumpType dump_db = eNoDump;
    Uint1 is_prot;
    FILE *outfp = NULL;
    CharPtr seqlocstr;
    Uint1 strand;
    Char buf[256] = { '\0' };
    Int2 rv;

    StringCpy(buf, "fastacmd ");
    StringNCat(buf, BlastGetVersionNumber(), sizeof(buf)-StringLen(buf));
    if (! GetArgs (buf, DIM(myargs), myargs)) {
	     return (1);
    }

    if( !ErrSetLogfile ("stderr", ELOG_APPEND) ) {
	     exit(1);
    }

    database = myargs[0].strvalue;
    if (!StringICmp(myargs[1].strvalue, "T"))
        is_prot = READDB_DB_IS_PROT;
    else if (!StringICmp(myargs[1].strvalue, "F"))
        is_prot = READDB_DB_IS_NUC;
    else
        is_prot = READDB_DB_UNKNOWN;

    searchstr     = myargs[2].strvalue;
    batchfile     = myargs[3].strvalue;
    dupl          = myargs[4].intvalue;
    linelen       = myargs[5].intvalue;
    target        = myargs[6].intvalue;
    use_ctrlAs    = myargs[8].intvalue;
    dump_db       = myargs[9].intvalue;
    seqlocstr     = myargs[10].strvalue;
    strand        = myargs[11].intvalue;
    taxonomy_info = myargs[12].intvalue;
    dbinfo_only   = myargs[13].intvalue;
    pig           = myargs[14].intvalue == 0 ? PIG_NONE : myargs[14].intvalue;

    if ((outfp = FileOpen(myargs[7].strvalue, "w")) == NULL) {
        ErrPostEx(SEV_ERROR, 0, 0,"Could not open %s", myargs[7].strvalue);
        return 1;
    }

    rv = Fastacmd_Search_ex (searchstr, database, is_prot, batchfile, dupl,
            linelen, outfp, target, use_ctrlAs, dump_db, seqlocstr, strand,
            taxonomy_info, dbinfo_only, pig);

    FileClose(outfp);

    return rv;
}
